### ISSUE TITLE
ENH: Productionize AWS Lambda Workflow

### ISSUE THREAD

**Issue by csmcallister** [DataKind-DC/capital-nature-ingest#213]
## Expected Behavior

Now that we can deploy the scrapers to AWS Lambda, we can develop a fully automated solution by adding the following components:

 - [ ] A *lambda function* that triggers on S3 PUT Object events to send pre-signed urls of the s3 object to the capital nature stakeholders. We could use SNS or maybe some Python smtp library for the email service.
 - [ ] Another *lambda function* that triggers on the event of the log file being put into S3 in order to create bug issues in this repo.
 - [ ] A continuous delivery option within CircleCI for optional one-click deploy.
 - [x] Make the current lambda that performs the web-scraping idempotent by only putting the results into S3 if an object with that key does not already exist. This is important because, even though a lambda can be scheduled once a month like a cron job, it might actually execute more than once due to retries that are beyond the control of the lambda function itself (see [this](https://aws.amazon.com/premiumsupport/knowledge-center/lambda-function-idempotent/) for more).

## Possible Solution

In order to send the reports to Capital Nature in a single email that's triggered by a single S3 PUT:Object event, we'll need to do one of two things:
1. Rewrite parts of `events/utils/reports.py` to generate a single excel file that contains the 4 csvs they need as 4 separate sheets. That way we there is only one lambda execution and one email sent.
2. Leave `events/utils/reports.py` as-is, which will sequentially put the csvs into S3. For the lambda that is triggered on the put object events, parse the event message to find the object key that triggered the lambda. If the key is that of the last file that is put, then we know the others are there and we can get those along with pre-signed urls to send in them email. 

For CI/CD we should probably swap out the `circleci/python:3.7.3` base image for the AWS Linux 2 image we're using for local dev. Then steps should be lint, schema tests, build lambda(s), deploy cdk. We might need to use [artifacts](https://circleci.com/docs/2.0/artifacts/) for saving changesets in `cdk.out/`. Alternatively, we might want to checkout the [Severless Application Model](https://aws.amazon.com/serverless/sam/), which might be more mature for this use case.

For the lambda that adds bugs to this repo, we can use the [api](https://developer.github.com/v3/issues/#create-an-issue) in response to S3 Put:Object events. Need to be careful with rate limiting. Might also want to consider creating a separate bucket for logs so that the lambda doesn't fire unnecessarily (e.g. when the reports get put into the bucket).

## Context
Adding these pieces will fully automate the workflow, including the creation of issues for bugs.

**Comment by csmcallister** [comment]
@tuffacton I tried to outline a serverless solution here. I'd love to hear your thoughts/comments on the approach. Particurlalrly any gotchas with using SNS and pre-signed s3 urls, as I haven't used those before. Also curious about your experience with AWS Serverless Application Model (SAM), because using that might be better than using the CDK in our case? Thanks! 👏

**Comment by tuffacton** [comment]
@csmcallister I think I grok. When I first approached the scraper solution that you all have built I thought Lambda was the natural progression, but then I ran it and saw that on a t3.nano it took up a relatively long time, something like >5 mins. Lambda has far less compute capability and a timeout of 6 mins and is really designed for doing things that are super small and fast like request/response for pre-signed URLs.

I think one way to do this that could be interesting is to have a Lambda function that also has the idempotent characteristics you've described, but is also more distributed. My ideal solution is that we have one Lambda running on a Cloudwatch cron that calls many Lambda functions wrapped around the various scrapers. This way the overall process takes only the time of the longest scrape, as opposed to all of them scraping serially. We can achieve idempotent characteristics by storing in a true datastore such as RDS or DynamoDB?

Probably worth an architecture deep-dive sometime if we want to discuss in a data-jam or video whiteboard session.

**Comment by csmcallister** [comment]
@tuffacton Lambdas timeout now at [15 minutes](https://aws.amazon.com/about-aws/whats-new/2018/10/aws-lambda-supports-functions-that-can-run-up-to-15-minutes/), but I've gotten execution time down to ~3 minutes locally and in lambda with multi-threading and a whopping 1GB memory (cloudwatch logs show we could do with way less memory, but I just set it to something nice and round for testing).

Ideally, we'd have a separate lambda for each scraper like you said and one master lambda to coordinate those workers and join their results. The only snag I see there is requiring every contributor to be good about declaring their scraper's dependencies and python version. It's doable with good docs and might even be what we want to do once we get to a stable state in terms of the set of scrapers we have.   

As for idempotency, a data store would be ideal, but I feel like the S3 mantra of write-once-read-many fits our use case pretty well and we can achieve idempotency with the lambda that does the scraping by only putting the results into S3 if an object with that key does not already exist but. That way there's no duplicate put events and therefore no duplicate emails sent by the lambda that triggers on that put event. Plus I feel like Dynamo or a RDBS will be more costly. 

At any rate, are you free sometime next week for a project night? @Frankie-Figz and I started meeting up at [Three Whistles](https://www.three-whistles.com/) in Clarendon/Courthouse if that's also convenient for you. Feel free to coordinate with us on Slack btw.
